Make FacetAccessType evaluate to SymbolicBindingType for type-of a BindSymbolicName

danakj opened the pull request:
The SymbolicBindingType refers to the type value that will be substituted in for the BindSymbolicName, but holds onto the EntityNameId from the BindSymbolicName instead of (or in addition to, for now) the instruction.

The EntityNameId will be used to look in the ScopeStack to find the witnesses either from the BindSymbolicName instruction, or other instructions that specify `impls` constraints against the EntityName.

This will allow us to have the `T` in `I(T)` resolve to a `.Self` reference in the type so that we get type equality with the binding's type: `T:! I(.Self)`.

danakj committed:
symbolic-binding-type

danakj commented:
Based on #6113, first commit is `symbolic-binding-type`

danakj commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -1680,6 +1680,21 @@ struct SymbolicBindingPattern {
 };
 
+// The constant value of a FacetAccessType for a symbolic facet value.
+struct SymbolicBindingType {
```
I do worry that the name `SymbolicBindingType` is a little misleading, as other things we name in this pattern are themselves types, but this is more of an "access type". I thought about `SymbolicBindingAccessType` or `SymbolicBindingAsType`. WDYT?

zygoloid replied:
I don't find this misleading -- I think this *is* always a type. (It's a symbolic type, but still a type, I think.)

danakj replied:
Ok, we can leave it then. FWIW I think the distinction I am hitting is that this evaluates to a type but isn't a type in this instruction form, and it loses the "access" part of the FacetAccessType name without moving any closer to the final evaluated type.

danakj commented on toolchain/check/type.cpp:
```diff
@@ -238,33 +239,86 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
+    return context.types().GetInstId(type_id);
+  }
+  return facet_value_id;
 }
```
There are these three operations which all want to move to the type of a facet value but in slightly different ways or with slightly different inputs. I want to figure out a way to consolidate them though...

danakj replied:
I have merged two of them, need to think more about how to shape things with the third still, but it's not fundamental to the PR.

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -2174,15 +2174,70 @@ auto TryEvalTypedInst<SemIR::BindSymbolicName>(EvalContext& eval_context,
+
+  // If we know which specific we're evaluating within and this is the type of
+  // an argument of that specific, its constant value refers to the type of the
+  // corresponding argument value.
```
```suggestion
  // If we know which specific we're evaluating within and this is the type component of
  // a facet parameter of the generic, its constant value refers to the type component of the
  // corresponding argument value of the specific.
```
I interpreted "the type of an argument" here as meaning typeof(argument), rather than as meaning that the argument is a facet and this is the type component of its (type, witness) pair, and was confused.

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -2174,15 +2174,70 @@ auto TryEvalTypedInst<SemIR::BindSymbolicName>(EvalContext& eval_context,
+      // Use the EntityNameId from the specific.
+      auto value_bind =
+          eval_context.insts().GetAs<SemIR::BindSymbolicName>(value_inst_id);
+      bind.entity_name_id =
```
Do we also need to update the type? I think the type in `value` could be more precise / substituted compared to the type in `inst`.

Oh, I see, the `Replace*WithConstantValue` below handles this. But it will also re-substitute into the `entity_name_id`, which seems error-prone -- we shouldn't be replacing that twice. I think it's not wrong per se currently, given that replacing the entity_name_id with its constant value happens to be idempotent, but it's a hazard for future development. I think it'd be more robust to replace all the fields here, and skip the `Replace*WithConstantValue` calls below along this path.

danakj replied:
Done

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -2174,15 +2174,70 @@ auto TryEvalTypedInst<SemIR::BindSymbolicName>(EvalContext& eval_context,
+  if (bind_name.bind_index().has_value()) {
+    if (auto value =
+            eval_context.GetCompileTimeBindValue(bind_name.bind_index());
+        value.has_value()) {
+      auto value_inst_id = eval_context.constant_values().GetInstId(value);
+      if (auto facet =
+              eval_context.insts().TryGetAs<SemIR::FacetValue>(value_inst_id)) {
+        // If the symbolic binding is replaced by a FacetValue (type + witness),
+        // evaluate to its type component.
+        return eval_context.constant_values().Get(facet->type_inst_id);
+      }
```
May be worth at least noting in a comment that this is effectively forming and evaluating a `FacetAccessType`? Given the small amount of duplicated code here I guess it's probably not worth attempting to share the implementation.

danakj replied:
Yeah that's a good point to add, done. I hadn't considered sharing code by constructing said FacetAccessType. But for now I have just made it clear it's doing that operation.

zygoloid commented on toolchain/check/eval_inst.cpp:
```diff
@@ -208,16 +208,32 @@ auto EvalConstantInst(Context& context, SemIR::FacetAccessType inst)
+        .type_id = SemIR::TypeType::TypeId,
+        .entity_name_id = bind_name->entity_name_id,
+        // TODO: This is to be removed, at which point replace NewSamePhase with
+        // NewAnyPhase.
```
`NewAnyPhase` would deduce the phase from the operands, which would give `Concrete`, whereas we want this to always be symbolic. I think we'd need a little more machinery here, or maybe we can just keep using `NewSamePhase`?

danakj replied:
Using Symbolic when the constant value does not change during eval with a specific ends up crashing in CHECK in generic. I will say something more in this TODO to explore that when we get there.

zygoloid commented on toolchain/check/import_ref.cpp:
```diff
@@ -757,6 +759,23 @@ static auto GetLocalNameId(ImportContext& context, SemIR::NameId import_name_id)
+// canonical EntityName (even if they are present there).
+//
+// No new work is generated by calling this function.
+static auto GetLocalEntityNameId(ImportRefResolver& resolver,
```
If this doesn't add new work, it should take the base class `ImportContext` not `ImportRefResolver`. (The difference is that the derived class has the ability to add new work.)

danakj replied:
Oh, thanks. Done

zygoloid commented on toolchain/check/member_access.cpp:
```diff
@@ -320,14 +320,8 @@ static auto LookupMemberNameInScope(Context& context, SemIR::LocId loc_id,
   if (auto assoc_type =
           context.types().TryGetAs<SemIR::AssociatedEntityType>(type_id)) {
     if (lookup_in_type_of_base) {
+      base_id = MoveToFacetValue(context, base_id);
```
I would expect a function named `MoveTo` to consume its input in some way. Could this be named something more like `GetCorrespondingFacetValue` or `GetFacetValueIfSymbolicBindingType`?

danakj replied:
I figured out how to get rid of this entirely in https://github.com/carbon-language/carbon-lang/pull/6146

zygoloid commented on toolchain/check/type.cpp:
```diff
@@ -238,33 +239,89 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
-  CARBON_CHECK(!context.insts().Is<SemIR::FacetValue>(inst_id));
+  CARBON_CHECK(!context.insts().Is<SemIR::SymbolicBindingType>(inst_id));
 
   return context.constant_values().GetConstantInstId(inst_id);
```
I wonder: can this function be rephrased as first mapping to the constant inst id, and then doing some postprocessing steps if needed? (Eg, mapping from a facet value -> its type component, or mapping from a symbolic binding type -> its facet value, depending on whether we want a facet-flavored result or a type-flavored result.) I guess it's not entirely clear to me what the contract is here, but I'd have expected that two instructions with the same constant value ought to produce the same canonicalized facet or type value, and going to the constant value first would guarantee that is the case.

danakj replied:
I have simplified this function and made things more crisp in https://github.com/carbon-language/carbon-lang/pull/6157

zygoloid commented on toolchain/check/type.cpp:
```diff
@@ -238,33 +239,89 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
+    CARBON_CHECK(context.types().Is<SemIR::FacetType>(type_id));
+    return context.types().GetInstId(type_id);
+  }
+}
```
I'm having trouble understanding what this function is doing. It looks like:

Given `inst_id` which is a facet value, perform some canonicalization step. Then:
If the type of `inst_id` is the type `type`, then return `inst_id`.
If the type of `inst_id` is a facet type other than `type`, then return the type of `inst_id`.

I find that behavior surprising -- that it sometimes returns the value of `inst_id`, and sometimes returns the type of `inst_id`. Is that the intent here?

danakj replied:
Yeah, this is what name_lookup wants, but it is now the only caller of this after some more changes for https://github.com/carbon-language/carbon-lang/pull/6157

I have a TODO over there atm asking why it's like this because things are even more confusing: specifically it wants to get the FacetType for a `facet as type` but it doesn't for other `facet`s. I am going to explain that before I decide to inline this function back over there or what.

danakj replied:
I now fully understand what's going on in name lookup (actually in member access), and I have clarified the situation there and moved this helper (in a different form) over to member access since it's really quite specific to that in https://github.com/carbon-language/carbon-lang/pull/6163

zygoloid commented on toolchain/check/type.h:
```diff
@@ -108,19 +108,40 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
-// For facet values, unwraps `FacetValue` instructions to get to an underlying
-// canonical type instruction.
-auto GetCanonicalizedFacetOrTypeValue(Context& context, SemIR::InstId inst_id)
+// If `preserve_facet_value` is false, the `FacetValue` instruction is unwrapped
+// to get to an underlying canonical type instruction. Otherwise, the output is
+// either the original instruction or a facet value instruction.
```
Maybe I'm misunderstanding something, but it looks to me like this comment doesn't match the implementation. Specifically, I think if we're given a `SymbolicBindingType`, we'll always produce a facet value regardless of `preserve_facet_value`, and won't produce a canoncial type instruction (which would be the `SymbolicBindingType`).

danakj replied:
This is gone now, thanks to https://github.com/carbon-language/carbon-lang/pull/6157

zygoloid commented on toolchain/check/testdata/facet/convert_facet_type_to_facet_value.carbon:
```diff
@@ -81,7 +81,7 @@ fn F() {
   Feed(Goat as (Animal & Climbs));
 
-  // CHECK:STDERR: fail_convert_multi_interface_facet_value_to_facet_value.carbon:[[@LINE+7]]:3: error: cannot convert type `Animal & Climbs` into type implementing `Eats` [ConversionFailureTypeToFacet]
+  // CHECK:STDERR: fail_convert_multi_interface_facet_value_to_facet_value.carbon:[[@LINE+7]]:3: error: cannot convert type `Animal & Climbs` that implements `Animal & Climbs` into type implementing `Eats` [ConversionFailureFacetToFacet]
```
This diagnostic doesn't look right to me.

danakj replied:
Yes, this is fixed with the changes for https://github.com/carbon-language/carbon-lang/pull/6157

zygoloid commented on toolchain/check/testdata/facet/convert_facet_value_value_to_generic_facet_value_value.carbon:
```diff
@@ -29,6 +29,13 @@ class Goat {}
 impl Goat as Animal {}
 
 fn Feed[Food:! Edible, T:! Eats(Food)](e: T, food: Food) {}
+// CHECK:STDERR: convert_facet_value_value_to_generic_facet_value_value.carbon:[[@LINE+7]]:64: error: cannot convert type `T` that implements `Animal` into type implementing `Eats(Food)` [ConversionFailureFacetToFacet]
```
Is this diagnostic expected? (If so I guess we need to rename this to `fail_todo_`)

danakj replied:
No, thanks for spotting it.

Interestingly, if I change `Eats` parameter from `Food:! type` to `Food:! Edible` the error goes away. But I still don't think the error is correct. The impls constrain the parameter given to `Eats` to be `Edible`, but they also should work when `Eats` only requires `type`.

danakj replied:
Details here: https://discord.com/channels/655572317891461132/941071822756143115/1423429988748623934

The tl;dr is that in deduce we substitute deduced arguments from previous bindings into future bindings. If your future binding has something like `U(T as type)`, so there's a `SymbolicBindingType` in its specific, then we try to substitute the deduced argument for `T` into it, which is a facet value. That can replace the `BindSymbolicName` along with its `EntityNameId`, which would then cause `FacetAccessType` to be rebuilt too, but it doesn't do anything to the `EntityNameId` in `SymbolicBindingType`, and we don't have a substitution for the as type part of that. So we have to do something like what eval is doing to `SymbolicBindingType`. But we can't modify a `SymbolicBindingType` in the same way, so we have to make a new `FacetAccessType` with the replacement that gets evaluated back into a `SymbolicBindingType`.

geoffromer replied:
Just to confirm, this was fixed, right?

danakj replied:
Yep, this is fixed by: https://github.com/carbon-language/carbon-lang/pull/6115/files#diff-e75f19a6bc525cf4282ee3697a2dc7722b2163694f69643dc1ba50ad1587b988R427-R446

danakj reviewed (commented):
uploading progress so far

danakj commented:
(not ready for more review yet)

danakj committed:
name-lookup-facettype

danakj commented:
I addressed a bunch of the comments here in other PRs, so I have rebased this on top of them (most of which have already landed). The first commit is still `symbolic-binding-type`. It's now based on #6163.

danakj commented:
I've marked this for review again by @zygoloid but @geoffromer would you mind to pick up the review from where he had left off?

geoffromer commented on toolchain/check/import_ref.cpp:
```diff
@@ -761,6 +763,23 @@ static auto GetLocalNameId(ImportContext& context, SemIR::NameId import_name_id)
+// canonical EntityName (even if they are present there).
+//
+// No new work is generated by calling this function.
+static auto GetLocalEntityNameId(ImportContext& context,
```
It looks like this function is only used with the names of symbolic bindings, and the implementation seems to assume that as well (e.g. using `AddSymbolicBindingName`, and omitting the `NameScopeId`). Should the name and comment explicitly restrict it to those names?

danakj replied:
Yeah, most places don't import EntityNameId at all. We only import them for symbolic bindings. Putting something in the name here won't hurt.

Though note that the NameScopeId is not part of the canonical EntityNameId (see the comment on the function). So the other fields are not imported because we import canonical values.

geoffromer reviewed (commented):
First batch of comments; I'm still reviewing the testdata.

danakj committed:
symbolic-name

geoffromer commented on toolchain/check/testdata/facet/convert_facet_value_value_to_generic_facet_value_value.carbon:
```diff
@@ -29,7 +29,7 @@ class Goat {}
 
 fn Feed[Food:! Edible, T:! Eats(Food)](e: T, food: Food) {}
-fn HandleAnimal[T:! Animal, Food:! Edible](a: T, food: Food) { Feed(a, food); }
+fn HandleAnimal[A:! Animal, Food:! Edible](a: A, food: Food) { Feed(a, food); }
```
Is this edit just to more clearly distinguish it from `Feed`'s `T` in the SemIR, or is there a deeper reason?

danakj replied:
yeah this made debugging easier so i left it in

geoffromer commented on toolchain/check/testdata/impl/use_assoc_const.carbon:
```diff
@@ -1772,12 +1778,16 @@ fn F() {
+// CHECK:STDOUT:     %.loc13_29.1: %facet_type = converted constants.%as_type.2a6e07.2, %facet_value.loc13_29.1 [symbolic = %facet_value.loc13_15.3 (constants.%facet_value)]
+// CHECK:STDOUT:     %as_type.loc13_29.2: type = facet_access_type constants.%impl.elem0.82c3d7.2 [symbolic = %as_type.loc12_35.1 (constants.%as_type.2a6e07.2)]
+// CHECK:STDOUT:     %facet_value.loc13_29.2: %facet_type = facet_value constants.%as_type.2a6e07.2, (constants.%I.lookup_impl_witness, constants.%Destroy.lookup_impl_witness) [symbolic = %facet_value.loc13_15.3 (constants.%facet_value)]
+// CHECK:STDOUT:     %.loc13_29.2: %facet_type = converted constants.%as_type.2a6e07.2, %facet_value.loc13_29.2 [symbolic = %facet_value.loc13_15.3 (constants.%facet_value)]
```
Are these changes expected? I haven't seen anything like them in the other files so far.

danakj replied:
Ah good catch, thank you. This was the shortcut in convert being changed to handle SymbolicBindingType *instead of* FacetAccessType. Like in other places, I have now built it on top of GetCanonicalFacetOrTypeValue so we also correctly shortcut when the FacetAccessType was on an associated constant rather than a binding.

geoffromer commented on toolchain/check/import_ref.cpp:
```diff
@@ -761,6 +763,23 @@ static auto GetLocalNameId(ImportContext& context, SemIR::NameId import_name_id)
   return import_name_id;
 }
 
+// Returns the id for a local EntityName from an imported one, preserving only
```
```suggestion
// Returns the id for a local symbolic EntityName from an imported one, preserving only
```

danakj replied:
Done

danakj committed:
review

danakj committed:
shortcut

danakj reviewed (commented):
PTAL
